zebra-network/src/peer_set/candidate_set/tests.rs

```diff
@@ -0,0 +1,124 @@
+//! [`CandidateSet`] tests.
+
+use std::{
+    future,
+    net::{IpAddr, SocketAddr},
+    sync::{Arc, Mutex},
+    task::{Context, Poll},
+};
+
+use chrono::{Duration, Utc};
+use tower::Service;
+use tracing::Level;
+
+use super::{AddressBook, BoxError, CandidateSet, MetaAddr, Request, Response};
+use crate::{protocol::types::PeerServices, Config};
+
+/// Test if gossiped peers' reported last seen time is properly normalized.
+///
+/// This is a security issue if this test fails, because a malicious peer can force the local node
+/// to use the peers it wants to, by reporting a last seen time that's in the future.
+///
+/// Note that this issue can also occur due to skewed clocks of different machines.
+///
+/// For more information, see the respective [issue] on GitHub.
+///
+/// [issue]: https://github.com/ZcashFoundation/zebra/issues/1871
+#[tokio::test]
+async fn gossiped_peers_from_a_single_peer_doesnt_get_full_precedence() {
+    // Use port numbers to mark peers to distinguish malicious ones from normal ones in the test
+    const MALICIOUS_PEER_PORT: u16 = 20_666;
+    const NORMAL_PEER_PORT: u16 = 20_111;
+    const SEED_PEER_PORT: u16 = 20_000;
+
+    let now = Utc::now();
+
+    let malicious_peer = MetaAddr::new_responded(
+        &SocketAddr::from((IpAddr::from([192, 168, 1, 1]), SEED_PEER_PORT)),
+        &PeerServices::NODE_NETWORK,
+    );
+
+    let normal_peer = MetaAddr::new_responded(
+        &SocketAddr::from((IpAddr::from([192, 168, 1, 2]), SEED_PEER_PORT)),
+        &PeerServices::NODE_NETWORK,
+    );
+
+    let mut address_book = AddressBook::new(&Config::default(), span!(Level::WARN, "Test span"));
+
+    address_book.extend(vec![normal_peer, malicious_peer]);
+
+    let peers_gossiped_by_normal_peer = vec![
+        MetaAddr::new_gossiped_meta_addr(
+            SocketAddr::from((IpAddr::from([192, 168, 1, 101]), NORMAL_PEER_PORT)),
+            PeerServices::NODE_NETWORK,
+            now - Duration::minutes(15),
+        ),
+        MetaAddr::new_gossiped_meta_addr(
+            SocketAddr::from((IpAddr::from([192, 168, 1, 102]), NORMAL_PEER_PORT)),
+            PeerServices::NODE_NETWORK,
+            now - Duration::minutes(45),
+        ),
+    ];
+
+    let peers_gossiped_by_malicious_peer = vec![
+        MetaAddr::new_gossiped_meta_addr(
+            SocketAddr::from((IpAddr::from([192, 168, 2, 101]), MALICIOUS_PEER_PORT)),
+            PeerServices::NODE_NETWORK,
+            now + Duration::minutes(30),
+        ),
+        MetaAddr::new_gossiped_meta_addr(
+            SocketAddr::from((IpAddr::from([192, 168, 2, 102]), MALICIOUS_PEER_PORT)),
+            PeerServices::NODE_NETWORK,
+            now + Duration::minutes(31),
+        ),
+    ];
```

Comment:
I think the test case data isn't quite testing the issue we're trying to prevent.

Here's an example of the security issue we're trying to prevent:
1. At time `T`, a malicious peer sends us gossiped addresses with a last seen time of `T+10_000`
2. At time `T+1`, an honest peer sends us gossiped addresses with a last seen time of `T+1`
3. Until `T+10_000`, we use the malicious addresses instead of the honest addresses

Here's an example of a valid peer priority that actually improves Zebra's performance:
1. At time `T`, an honest peer sends us gossiped addresses with a last seen time of `T`
2. At time `T+1`, an honest peer sends us gossiped addresses with a last seen time of `T-1000`
3. We use the addresses from the first honest peer